katsdpcontroller/product_config.py

```diff
@@ -31,63 +82,1196 @@ class _Sensor(ABC):
+                    options: Options,
+                    name: str,
+                    config: Mapping[str, Any],
+                    src_streams: Sequence['Stream'],
+                    sensors: Mapping[str, Any]) -> 'TiedArrayChannelisedVoltageStream':
+        return cls(
+            name, src_streams,
+            url=yarl.URL(config['url']),
+            n_channels_per_substream=sensors['n_chans_per_substream'],
+            spectra_per_heap=sensors['spectra_per_heap'],
+            bits_per_sample=sensors['beng_out_bits_per_sample'],
+            instrument_dev_name=config['instrument_dev_name']
+        )
+
+
+class SimTiedArrayChannelisedVoltageStream(TiedArrayChannelisedVoltageStreamBase):
+    """Simulated tied-array-channelised-voltage stream."""
+
+    stream_type: ClassVar[str] = 'sim.cbf.tied_array_channelised_voltage'
+    _valid_src_types: ClassVar[_ValidTypes] = {'sim.cbf.antenna_channelised_voltage'}
+
+    def __init__(self, name: str, src_streams: Sequence[Stream], *,
+                 n_endpoints: int,
+                 n_channels_per_substream: Optional[int] = None,
+                 spectra_per_heap: int) -> None:
+        acv = cast(AntennaChannelisedVoltageStream, src_streams[0])
+        if n_channels_per_substream is not None:
+            ncps = n_channels_per_substream
+        else:
+            ncps = acv.n_channels // n_endpoints
+        super().__init__(
+            name, src_streams,
+            n_endpoints=n_endpoints,
+            n_channels_per_substream=ncps,
+            spectra_per_heap=spectra_per_heap,
+            bits_per_sample=8
+        )
+
+    if TYPE_CHECKING:     # pragma: nocover
+        # Refine the return type for mypy
+        @property
+        def antenna_channelised_voltage(self) -> SimAntennaChannelisedVoltageStream: ...
+
+    @classmethod
+    def from_config(cls,
+                    options: Options,
+                    name: str,
+                    config: Mapping[str, Any],
+                    src_streams: Sequence['Stream'],
+                    sensors: Mapping[str, Any]) -> 'SimTiedArrayChannelisedVoltageStream':
+        return cls(
+            name, src_streams,
+            n_endpoints=config['n_endpoints'],
+            n_channels_per_substream=config.get('n_chans_per_substream'),
+            spectra_per_heap=config.get('spectra_per_heap', KATCBFSIM_SPECTRA_PER_HEAP)
+        )
+
+
+class VisStream(Stream):
+    """Instance of sdp.vis."""
+
+    stream_type: ClassVar[str] = 'sdp.vis'
+    _valid_src_types: ClassVar[_ValidTypes] = {
+        'cbf.baseline_correlation_products',
+        'sim.cbf.baseline_correlation_products'
+    }
+
+    def __init__(self, name: str, src_streams: Sequence[Stream], *,
+                 int_time: float,
+                 output_channels: Optional[Tuple[int, int]] = None,
+                 continuum_factor: int,
+                 excise: bool,
+                 archive: bool,
+                 n_servers: int) -> None:
```

Comment:
Could you call this n_ingest_servers just to be clear ?

Reply:
I'm reluctant because at some point I will probably want to give CalStream the same property and maybe even share some code to validate that the number of servers divides into the number of channels, and that's easier if they have a common attribute.

Reply:
Fair enough - perhaps n_instances ? I just don't like n_servers by itself since it carries hints of a mapping to physical hardware which is not necessarily true. Although we are getting dangerously close to the shed with this one :)

Reply:
I see your point, but the term "server" is also used in the command-line interfaces to katsdpingest, katcbfsim, and katsdpcal, so I'd rather stay consistent with that (and trying to change all of those to use "instances" and dealing with the compatibility issues doesn't seem worth the trouble).

Reply:
OK